[VPlan] Add BranchOnTwoConds, use for early exit plans. (#172750)
llvm/llvm-project: llvm/docs/Vectorizers.rst

When vectorizing a loop with a single early exit, the loop blocks following the
early exit are predicated and the vector loop will always exit via the latch.

> It is terminated by a BranchOnTwoConds VPInstruction, which takes both the early

Reviewer: ```suggestion
The loop terminates with a single BranchOnTwoConds VPInstruction, which takes both the early
```

and latch exiting conditions. If the early exiting condition is true,
BranchOnToConds exits to an intermediate block (``vector.early.exit`` below).
This intermediate block is responsible for calculating any exit values of
loop-defined variables that are used in the early exit block. If the latch
exiting condition is true, exits to the ``middle.block`` which selects between

> the exit block from the latch or the scalar remainder loop. Otherwise continues

Author: updated ,thanks

> executing in the region header.

Reviewer: ```suggestion
executing the loop by jumping back to the region header.
```

Author: done thanks

When vectorizing a loop with a single early exit, the loop blocks following the
early exit are predicated and the vector loop will always exit via the latch.

> The loop terminates with a single BranchOnTwoConds VPInstruction, which takes

Reviewer: @fhahn 's original proposal was to support an arbitrary number of conditions using a single `BranchOnMultiCond`, but having currently only a single instance which exercises two conditions it was trimmed down to two, expecting to extend it to two-or-more when the need (i.e., multiple early exits) arises, along with test coverage.
